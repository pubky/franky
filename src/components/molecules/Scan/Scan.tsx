'use client';

import Image from 'next/image';
import { QRCodeSVG } from 'qrcode.react';
import { useEffect, useRef, useState } from 'react';
import { useRouter } from 'next/navigation';

import * as Atoms from '@/atoms';
import * as Libs from '@/libs';
import * as Molecules from '@/molecules';
import * as Config from '@/config';
import * as Core from '@/core';
import * as App from '@/app';

export const ScanContent = () => {
  const [url, setUrl] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const retryCountRef = useRef(0);
  const isMountedRef = useRef(true);
  const activeRequestRef = useRef<symbol | null>(null);
  const isGeneratingRef = useRef(false);
  const MAX_RETRY_ATTEMPTS = 3;

  const isIOS = typeof navigator !== 'undefined' && /iPhone|iPad|iPod/i.test(navigator.userAgent);
  const fallbackUrl = isIOS ? Config.APP_STORE_URL : Config.PLAY_STORE_URL;

  const fetchUrl = async (options?: { viaRetry?: boolean }) => {
    const requestId = Symbol('fetchUrl');
    activeRequestRef.current = requestId;
    isGeneratingRef.current = true;
    if (!options?.viaRetry) {
      setIsLoading(true);
      setUrl('');
    }

    let willRetry = false;

    try {
      const data = await Core.AuthController.getAuthUrl();
      if (!data) return;

      const { url: generatedUrl, promise } = data;

      // Always attach handlers to avoid unhandled rejections even if unmounted
      promise
        ?.then(async (publicKey) => {
          // Ignore if unmounted or superseded
          if (activeRequestRef.current !== requestId || !isMountedRef.current) return;
          try {
            await Core.AuthController.loginWithAuthUrl({ publicKey });
          } catch (error) {
            Libs.Logger.error('Failed to login with auth URL:', error);
            if (!isMountedRef.current) return;
            Molecules.toast({
              title: 'Authorization failed',
              description: 'Unable to complete authorization with Pubky Ring. Please try again.',
            });
            if (activeRequestRef.current === requestId) {
              void fetchUrl();
            }
          }
        })
        .catch((error: unknown) => {
          // Rejected authorization or transport failure
          Libs.Logger.error('Authorization promise rejected:', error);
          if (!isMountedRef.current) return;
          Molecules.toast({
            title: 'Authorization was not completed',
            description: 'The signer did not complete authorization. Please try again.',
          });
          if (activeRequestRef.current === requestId) {
            void fetchUrl();
          }
        });

      // Guard against late responses from previous calls
      if (activeRequestRef.current !== requestId || !isMountedRef.current) return;

      if (generatedUrl) setUrl(generatedUrl);
      retryCountRef.current = 0;
    } catch (error) {
      retryCountRef.current += 1;
      const attempts = retryCountRef.current;
      Libs.Logger.error(`Failed to generate auth URL (attempt ${attempts} of ${MAX_RETRY_ATTEMPTS}):`, error);

      if (attempts < MAX_RETRY_ATTEMPTS) {
        willRetry = true;
        // bounded backoff: 250ms, 500ms
        const delayMs = Math.min(1000, 250 * attempts);
        await new Promise((resolve) => setTimeout(resolve, delayMs));
        await fetchUrl({ viaRetry: true });
      } else if (isMountedRef.current) {
        Molecules.toast({
          title: 'QR code generation failed',
          description: 'Unable to generate auth QR code. Please refresh and try again.',
        });
      }
    } finally {
      // Only clear loading if we are not immediately retrying and this is the latest request
      if (!willRetry && activeRequestRef.current === requestId) {
        isGeneratingRef.current = false;
        if (isMountedRef.current) setIsLoading(false);
      }
    }
  };

  const copyAuthUrlToClipboard = async () => {
    if (!url) return;

    try {
      await Libs.copyToClipboard({ text: url });
    } catch (error) {
      Libs.Logger.error('Failed to copy auth URL to clipboard:', error);
    }
  };

  const handleMobileAuth = async () => {
    if (isLoading || isGeneratingRef.current) return;

    if (!url) {
      if (activeRequestRef.current) return;
      void fetchUrl();
      return;
    }

    await copyAuthUrlToClipboard();

    const deeplink = `pubkyring://${url}`;

    try {
      const openedWindow = window.open(deeplink, '_blank');

      if (!openedWindow) {
        window.location.href = deeplink;
        return;
      }

      setTimeout(() => {
        try {
          openedWindow.location.href = fallbackUrl;
        } catch (error) {
          Libs.Logger.error('Failed to redirect to store after deeplink attempt:', error);
          window.location.href = fallbackUrl;
        }
      }, 2000);
    } catch (error) {
      Libs.Logger.error('Failed to open Pubky Ring deeplink:', error);
      Molecules.toast({
        title: 'Unable to link to signer application Pubky Ring',
        description: 'Please try again.',
      });
      window.location.href = fallbackUrl;
    }
  };

  useEffect(() => {
    isMountedRef.current = true;
    fetchUrl();
    return () => {
      isMountedRef.current = false;
      activeRequestRef.current = null;
      isGeneratingRef.current = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      {/** Desktop view */}
      <Atoms.Container size="container" className="hidden md:flex">
        <ScanHeader isMobile={false} />
        <Molecules.ContentCard layout="column">
          <Atoms.Container className="items-center justify-center">
            <div className="bg-foreground rounded-lg p-4 w-[220px] h-[220px] flex items-center justify-center">
              {isLoading || !url ? (
                <Atoms.Container className="items-center gap-2">
                  <Libs.Loader2 className="h-8 w-8 animate-spin text-background" />
                  <Atoms.Typography as="small" size="sm" className="text-background">
                    Generating QR Code...
                  </Atoms.Typography>
                </Atoms.Container>
              ) : (
                <QRCodeSVG value={url} size={220} />
              )}
            </div>
          </Atoms.Container>
        </Molecules.ContentCard>
      </Atoms.Container>

      {/** Mobile view */}
      <Atoms.Container size="container" className="md:hidden">
        <ScanHeader isMobile={true} />
        <Molecules.ContentCard layout="column">
          <Atoms.Container className="flex-col lg:flex-row gap-12 items-center justify-center">
            <Image src="/images/logo-pubky-ring.svg" alt="Pubky Ring" width={137} height={30} />
            <Atoms.Button
              className="w-full h-[60px] rounded-full"
              size="lg"
              onClick={handleMobileAuth}
              disabled={isLoading || !url}
            >
              {isLoading ? (
                <>
                  <Libs.Loader2 className="mr-2 h-4 w-4 animate-spin" />
                  Generating...
                </>
              ) : (
                <>
                  <Libs.Key className="mr-2 h-4 w-4" />
                  Authorize with Pubky Ring
                </>
              )}
            </Atoms.Button>
          </Atoms.Container>
        </Molecules.ContentCard>
      </Atoms.Container>
    </>
  );
};

export const ScanFooter = () => {
  return (
    <Atoms.FooterLinks className="py-6">
      Use{' '}
      <Atoms.Link href={Config.PUBKY_RING_URL} target="_blank">
        Pubky Ring
      </Atoms.Link>{' '}
      or any other{' '}
      <Atoms.Link href={Config.PUBKY_CORE_URL} target="_blank">
        Pubky Core
      </Atoms.Link>
      â€“powered keychain.
    </Atoms.FooterLinks>
  );
};

export const ScanHeader = ({ isMobile }: { isMobile: boolean }) => {
  return (
    <Atoms.PageHeader>
      <Molecules.PageTitle size="large">
        {isMobile ? (
          <>
            Tap to <span className="text-brand">Authorize.</span>
          </>
        ) : (
          <>
            Scan <span className="text-brand">QR Code.</span>
          </>
        )}
      </Molecules.PageTitle>
      <Atoms.PageSubtitle>Open Pubky Ring, create a pubky, and scan the QR.</Atoms.PageSubtitle>
    </Atoms.PageHeader>
  );
};

export const ScanNavigation = () => {
  const router = useRouter();

  const onHandleBackButton = () => {
    router.push(App.ONBOARDING_ROUTES.INSTALL);
  };

  return (
    <Molecules.ButtonsNavigation
      continueButtonDisabled={true}
      hiddenContinueButton={true}
      onHandleBackButton={onHandleBackButton}
    />
  );
};
