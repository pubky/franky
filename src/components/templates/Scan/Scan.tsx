import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';

export function Scan() {
  return (
    <Atoms.Container size="container">
      <Molecules.ScanContent />
      <Molecules.ScanFooter />
      <Molecules.ScanNavigation />
    </Atoms.Container>
  );
}
