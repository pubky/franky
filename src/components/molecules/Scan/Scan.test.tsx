import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { ScanContent, ScanFooter, ScanHeader, ScanNavigation } from './Scan';

// Mock Next.js router
const mockPush = vi.fn();
vi.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

// Mock Next.js Image
vi.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, width, height }: { src: string; alt: string; width: number; height: number }) => (
    // eslint-disable-next-line @next/next/no-img-element
    <img data-testid="next-image" src={src} alt={alt} width={width} height={height} />
  ),
}));

// Mock molecules
vi.mock('@/molecules', () => ({
  ResponsiveSection: ({ desktop, mobile }: { desktop: React.ReactNode; mobile: React.ReactNode }) => (
    <div data-testid="responsive-section">
      <div data-testid="desktop-content">{desktop}</div>
      <div data-testid="mobile-content">{mobile}</div>
    </div>
  ),
  ContentCard: ({ children, layout }: { children: React.ReactNode; layout?: string }) => (
    <div data-testid="content-card" data-layout={layout}>
      {children}
    </div>
  ),
  PageTitle: ({ children, size }: { children: React.ReactNode; size?: string }) => (
    <div data-testid="page-title" data-size={size}>
      {children}
    </div>
  ),
  ButtonsNavigation: ({
    onHandleBackButton,
    continueButtonDisabled,
    hiddenContinueButton,
  }: {
    onHandleBackButton: () => void;
    continueButtonDisabled?: boolean;
    hiddenContinueButton?: boolean;
  }) => (
    <div data-testid="buttons-navigation">
      <button data-testid="back-button" onClick={onHandleBackButton}>
        Back
      </button>
      {!hiddenContinueButton && (
        <button data-testid="continue-button" disabled={continueButtonDisabled}>
          Continue
        </button>
      )}
    </div>
  ),
}));

// Mock atoms
vi.mock('@/atoms', () => ({
  Container: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="container" className={className}>
      {children}
    </div>
  ),
  Button: ({ children, className, size }: { children: React.ReactNode; className?: string; size?: string }) => (
    <button data-testid="button" className={className} data-size={size}>
      {children}
    </button>
  ),
  FooterLinks: ({ children }: { children: React.ReactNode }) => <div data-testid="footer-links">{children}</div>,
  Link: ({ children, href, target }: { children: React.ReactNode; href: string; target?: string }) => (
    <a data-testid="link" href={href} target={target}>
      {children}
    </a>
  ),
  PageHeader: ({ children }: { children: React.ReactNode }) => <div data-testid="page-header">{children}</div>,
  PageSubtitle: ({ children }: { children: React.ReactNode }) => <div data-testid="page-subtitle">{children}</div>,
}));

describe('ScanContent', () => {
  it('renders desktop and mobile content containers', () => {
    render(<ScanContent />);

    const containers = screen.getAllByTestId('container');
    expect(containers.length).toBeGreaterThan(0);

    // Check for desktop container (hidden md:flex)
    const desktopContainer = containers.find((container) => container.className.includes('hidden md:flex'));
    expect(desktopContainer).toBeInTheDocument();

    // Check for mobile container (md:hidden)
    const mobileContainer = containers.find((container) => container.className.includes('md:hidden'));
    expect(mobileContainer).toBeInTheDocument();
  });

  it('renders QR code image in desktop version', () => {
    render(<ScanContent />);

    const images = screen.getAllByTestId('next-image');
    const qrImage = images.find((img) => img.getAttribute('src') === '/images/pubky-ring-qr-example.png');

    expect(qrImage).toBeInTheDocument();
    expect(qrImage).toHaveAttribute('alt', 'Pubky Ring');
    expect(qrImage).toHaveAttribute('width', '220');
    expect(qrImage).toHaveAttribute('height', '220');
  });

  it('renders logo and button in mobile version', () => {
    render(<ScanContent />);

    const images = screen.getAllByTestId('next-image');
    const logoImage = images.find((img) => img.getAttribute('src') === '/images/logo-pubky-ring.svg');

    expect(logoImage).toBeInTheDocument();
    expect(logoImage).toHaveAttribute('alt', 'Pubky Ring');
    expect(logoImage).toHaveAttribute('width', '137');
    expect(logoImage).toHaveAttribute('height', '30');

    expect(screen.getByTestId('button')).toBeInTheDocument();
    expect(screen.getByText('Authorize with Pubky Ring')).toBeInTheDocument();
  });

  it('renders content cards with column layout', () => {
    render(<ScanContent />);

    const contentCards = screen.getAllByTestId('content-card');
    contentCards.forEach((card) => {
      expect(card).toHaveAttribute('data-layout', 'column');
    });
  });
});

describe('ScanFooter', () => {
  it('renders footer with links', () => {
    render(<ScanFooter />);

    expect(screen.getByTestId('footer-links')).toBeInTheDocument();
    expect(screen.getByText('Use', { exact: false })).toBeInTheDocument();
    expect(screen.getByText('or any other', { exact: false })).toBeInTheDocument();
    expect(screen.getByText('â€“powered keychain.', { exact: false })).toBeInTheDocument();
  });

  it('renders Pubky Ring link', () => {
    render(<ScanFooter />);

    const links = screen.getAllByTestId('link');
    const pubkyRingLink = links[0];

    expect(pubkyRingLink).toHaveAttribute('href', 'https://www.pubkyring.to/');
    expect(pubkyRingLink).toHaveAttribute('target', '_blank');
    expect(pubkyRingLink).toHaveTextContent('Pubky Ring');
  });

  it('renders Pubky Core link', () => {
    render(<ScanFooter />);

    const links = screen.getAllByTestId('link');
    const pubkyCoreLink = links[1];

    expect(pubkyCoreLink).toHaveAttribute('href', 'https://pubky.org');
    expect(pubkyCoreLink).toHaveAttribute('target', '_blank');
    expect(pubkyCoreLink).toHaveTextContent('Pubky Core');
  });
});

describe('ScanHeader', () => {
  it('renders mobile header correctly', () => {
    render(<ScanHeader isMobile={true} />);

    expect(screen.getByTestId('page-header')).toBeInTheDocument();
    expect(screen.getByTestId('page-title')).toBeInTheDocument();
    expect(screen.getByTestId('page-subtitle')).toBeInTheDocument();

    const title = screen.getByTestId('page-title');
    expect(title).toHaveAttribute('data-size', 'medium');
    expect(title).toHaveTextContent('Tap to Authorize.');
  });

  it('renders desktop header correctly', () => {
    render(<ScanHeader isMobile={false} />);

    expect(screen.getByTestId('page-header')).toBeInTheDocument();
    expect(screen.getByTestId('page-title')).toBeInTheDocument();
    expect(screen.getByTestId('page-subtitle')).toBeInTheDocument();

    const title = screen.getByTestId('page-title');
    expect(title).toHaveAttribute('data-size', 'medium');
    expect(title).toHaveTextContent('Scan QR Code.');
  });

  it('renders correct subtitle', () => {
    render(<ScanHeader isMobile={true} />);

    expect(screen.getByText('Open Pubky Ring, create a pubky, and scan the QR.')).toBeInTheDocument();
  });

  it('contains brand-styled text', () => {
    render(<ScanHeader isMobile={false} />);

    const title = screen.getByTestId('page-title');
    expect(title.textContent).toContain('QR Code');
  });
});

describe('ScanNavigation', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders buttons navigation with correct configuration', () => {
    render(<ScanNavigation />);

    expect(screen.getByTestId('buttons-navigation')).toBeInTheDocument();
    expect(screen.getByTestId('back-button')).toBeInTheDocument();
    expect(screen.queryByTestId('continue-button')).not.toBeInTheDocument(); // hidden
  });

  it('handles back button click', () => {
    render(<ScanNavigation />);

    const backButton = screen.getByTestId('back-button');
    fireEvent.click(backButton);

    expect(mockPush).toHaveBeenCalledWith('/onboarding/install');
  });

  it('hides continue button and disables it', () => {
    render(<ScanNavigation />);

    // Continue button should be hidden based on hiddenContinueButton prop
    expect(screen.queryByTestId('continue-button')).not.toBeInTheDocument();
  });
});
