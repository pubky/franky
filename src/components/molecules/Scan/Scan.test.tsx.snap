// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Scan Components - Snapshots > ScanContent - Snapshots > matches snapshot for default ScanContent 1`] = `
<div
  class="hidden md:flex"
  data-testid="container"
>
  <div
    data-testid="page-header"
  >
    <div
      data-size="large"
      data-testid="page-title"
    >
      Scan 
      <span
        class="text-brand"
      >
        QR Code.
      </span>
    </div>
    <div
      data-testid="page-subtitle"
    >
      Open Pubky Ring, create a pubky, and scan the QR.
    </div>
  </div>
  <div
    data-layout="column"
    data-testid="content-card"
  >
    <div
      class="items-center justify-center"
      data-testid="container"
    >
      <div
        class="bg-foreground rounded-lg p-4 w-[220px] h-[220px] flex items-center justify-center"
      >
        <div
          class="items-center gap-2"
          data-testid="container"
        >
          <svg
            aria-hidden="true"
            class="h-8 w-8 animate-spin text-background"
            data-testid="loader-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 12a9 9 0 1 1-6.219-8.56"
            />
          </svg>
          <small
            class="text-background"
            data-size="sm"
            data-testid="typography"
          >
            Generating QR Code...
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Scan Components - Snapshots > ScanFooter - Snapshots > matches snapshot for default ScanFooter 1`] = `
<div
  data-testid="footer-links"
>
  Use
   
  <a
    data-testid="link"
    href="https://pubkyring.app/"
    target="_blank"
  >
    Pubky Ring
  </a>
   
  or any other
   
  <a
    data-testid="link"
    href="https://pubky.org"
    target="_blank"
  >
    Pubky Core
  </a>
  â€“powered keychain.
</div>
`;

exports[`Scan Components - Snapshots > ScanHeader - Snapshots > matches snapshot for desktop ScanHeader 1`] = `
<div
  data-testid="page-header"
>
  <div
    data-size="large"
    data-testid="page-title"
  >
    Scan 
    <span
      class="text-brand"
    >
      QR Code.
    </span>
  </div>
  <div
    data-testid="page-subtitle"
  >
    Open Pubky Ring, create a pubky, and scan the QR.
  </div>
</div>
`;

exports[`Scan Components - Snapshots > ScanHeader - Snapshots > matches snapshot for mobile ScanHeader 1`] = `
<div
  data-testid="page-header"
>
  <div
    data-size="large"
    data-testid="page-title"
  >
    Tap to 
    <span
      class="text-brand"
    >
      Authorize.
    </span>
  </div>
  <div
    data-testid="page-subtitle"
  >
    Open Pubky Ring, create a pubky, and scan the QR.
  </div>
</div>
`;

exports[`Scan Components - Snapshots > ScanNavigation - Snapshots > matches snapshot for default ScanNavigation 1`] = `
<div
  data-testid="buttons-navigation"
>
  <button
    data-testid="back-button"
  >
    Back
  </button>
</div>
`;
